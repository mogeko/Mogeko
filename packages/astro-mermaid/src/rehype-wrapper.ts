import { visit } from "unist-util-visit";
import { h } from "hastscript";
import ClassList from "hast-util-class-list";
import type { Element } from "hast";

export function wrapper({ properties = {}, tagName = "p" }: Options = {}) {
  return function (tree: Element) {
    visit(tree, { tagName: "pre" }, (node, index = 0, parents) => {
      if (ClassList(node).contains("mermaid")) {
        if (parents) {
          tree.children[index] = h(tagName, properties, [node]);
        }
      }
    });
  };
}

type Options = Partial<Pick<Element, "tagName" | "properties">>;
