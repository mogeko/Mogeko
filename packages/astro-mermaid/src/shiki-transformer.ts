import { addClassToHast, type ShikijiTransformer } from "shikiji";

export function addMermaidClass(): ShikijiTransformer {
  return {
    name: "add-mermaid-class",
    pre(this, hast) {
      if (this.options.lang === "mermaid") {
        addClassToHast(hast, "mermaid");
      }

      return hast;
    },
  };
}
