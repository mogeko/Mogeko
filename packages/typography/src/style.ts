import type { CSSObject } from "@emotion/serialize";

const round = (num: number) => {
  return num
    .toFixed(7)
    .replace(/(\.[0-9]+?)0+$/, "$1")
    .replace(/\.0$/, "");
};
const rem = (px: number) => `${round(px / 16)}rem`;
const em = (px: number, base: number) => `${round(px / base)}em`;
const hsl = (val: string) => `hsl(${val})`;

export const styles: Styles = {
  DEFAULT: {
    a: {
      "--tw-shadow": "inset 0 -1px 0 0",
      "--tw-shadow-colored": "inset 0 -1px 0 0 var(--tw-shadow-color)",
      boxShadow:
        "var(--tw-ring-offset-shadow, 0 0 #0000), var(--tw-ring-shadow, 0 0 #0000), var(--tw-shadow)",
      position: "relative",
      "&:hover": {
        "--tw-shadow": `inset 0 ${em(-8, 16)} 0 0 var(--tw-shadow-color)`,
        "--tw-shadow-color": hsl("var(--primary)"),
        "--tw-shadow-colored": `inset 0 ${em(-8, 16)} 0 0 var(--tw-shadow-color)`,
      },
    },
    blockquote: {
      borderColor: hsl("var(--primary)"),
      borderLeftWidth: "5px",
      fontStyle: "italic",
      lineHeight: rem(28),
      margin: `0 ${rem(28)} ${rem(28)} ${rem(24)}`,
      opacity: 0.8,
      paddingLeft: rem(20),
    },
    code: {
      fontFamily: `ui-monospace, SFMono-Regular, Menlo, Monaco, Consolas, "Liberation Mono", "Courier New", monospace`,
      "&:not(:where(.astro-code *))": {
        backgroundColor: hsl("var(--muted)"),
        borderRadius: rem(4),
        fontSize: rem(14),
        fontWeight: 600,
        lineHeight: rem(20),
        padding: `${rem(3.2)} ${rem(4.8)}`,
        position: "relative",
      },
    },
    "h1, h2, h3, h4, h5, h6": {
      letterSpacing: em(-0.9, 36),
      scrollMargin: rem(80),
    },
    h1: {
      fontSize: rem(36),
      fontWeight: 900,
      lineHeight: rem(40),
      "&:not(:first-child)": {
        margin: `${rem(56)} 0 ${rem(28)}`,
      },
    },
    h2: {
      fontSize: rem(30),
      fontWeight: 800,
      lineHeight: rem(36),
      paddingBottom: rem(8),
      "&:not(:first-child)": {
        margin: `${rem(48)} 0 ${rem(24)}`,
      },
    },
    h3: {
      fontSize: rem(24),
      fontWeight: 700,
      lineHeight: rem(32),
      "&:not(:first-child)": {
        margin: `${rem(40)} 0 ${rem(20)}`,
      },
    },
    h4: {
      fontSize: rem(20),
      fontWeight: 600,
      lineHeight: rem(28),
      "&:not(:first-child)": {
        margin: `${rem(32)} 0 ${rem(16)}`,
      },
    },
    img: {
      borderRadius: "calc(var(--radius) - 4px)",
    },
    "ol, ul": {
      fontWeight: 500,
      lineHeight: 1.1,
      marginTop: rem(16),
      "& > li": {
        marginBottom: rem(16),
        paddingLeft: rem(12),
        "&::marker": {
          color: hsl("var(--primary)"),
        },
      },
    },
    ol: {
      listStyleType: "decimal",
    },
    p: {
      letterSpacing: em(-0.16, 16),
      lineHeight: rem(28),
      "&:not(:last-child)": {
        marginTop: rem(24),
      },
    },
    ul: {
      listStyleType: "disc",
    },
  },
  lg: {
    h1: {
      fontSize: rem(48),
      lineHeight: 1,
    },
    h2: {
      fontSize: rem(36),
      lineHeight: rem(40),
    },
    h3: {
      fontSize: rem(30),
      lineHeight: rem(36),
    },
  },
};

export type Styles = {
  [key: string]: CSSObject;
  DEFAULT: CSSObject;
};
