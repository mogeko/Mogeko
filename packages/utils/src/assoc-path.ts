import { is } from "@/is";

/**
 * Makes a shallow clone of an object, setting or overriding the nodes required
 * to create the given path, and placing the specific value at the tail end of
 * that path. Note that this copies and flattens prototype properties onto the
 * new object as well. All non-primitive properties are copied by reference.
 *
 * @param path - the path to set
 * @param value - The new value
 * @param obj -  The object to clone
 * @return A new object equivalent to the original except along the specified path.
 *
 * @example
 * ```typescript
 * assocPath(["a", "b", "c"], 42, {a: {b: {c: 0}}}); //=> {a: {b: {c: 42}}}
 * // Any missing or non-object keys in path will be overridden
 * assocPath(["a", "b", "c"], 42, {a: 5});           //=> {a: {b: {c: 42}}}
 * // If the path includes a number, it will create an array
 * assocPath(["a", "b", 1, "c"], 42, {a: 5});        //=> {a: {b: [undefined, {c: 42}]}}
 * ```
 */
export function assocPath(path: Path, value: any, obj: Obj): Obj {
  if (path.length === 0) return value;

  const [idx, ...tail] = path as [Path[0], ...Path];
  const next = is(Object, obj[idx]) ? obj[idx] : is(Number, tail[0]) ? [] : {};

  return Object.assign(obj, { [idx]: assocPath(tail, value, next) });
}

type Path = (string | number)[];
type Obj = Record<string, any>;
