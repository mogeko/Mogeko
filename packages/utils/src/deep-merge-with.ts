import { isNil } from "@/is-nil";
import { is } from "@/is";

export function deepMergeWith(fn: (a: any, b: any) => any) {
  return (...objs: any[]) => {
    return objs.reduce((acc, obj) => {
      if (isNil(acc) || !is(Object, acc)) return obj;
      if (isNil(obj) || !is(Object, obj)) return acc;

      Object.keys(obj).forEach((key) => {
        if (!isNil(acc[key])) {
          if (is(Object, acc[key]) || is(Object, obj[key])) {
            acc[key] = deepMergeWith(fn)(acc[key], obj[key]);
          } else {
            acc[key] = fn(acc[key], obj[key]);
          }
        } else {
          acc[key] = obj[key];
        }
      });

      return acc;
    });
  };
}
