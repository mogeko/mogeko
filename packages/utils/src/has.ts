import { isNil } from "@/is-nil";

/**
 * Returns whether or not an object has an own property with the specified name
 *
 * @param key - The name of the property to check for.
 * @param obj - The object to query.
 * @return Whether the property exists.
 *
 * @example
 * ```typescript
 * has("a", { a: 1 }); //=> true
 * has("a", { b: 1 }); //=> false
 * has("a", null);     //=> false
 * ```
 */
export function has<K extends PropertyKey>(
  key: K,
  obj: unknown,
): obj is ObjectHavingSome<K> {
  return !isNil(obj) && Object.hasOwn(obj, key);
}

type ObjectHavingSome<Key extends PropertyKey> = {
  [K in Key]: { [P in K]: unknown };
}[Key];
