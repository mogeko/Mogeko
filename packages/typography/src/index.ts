import plugin from "tailwindcss/plugin";
import { styles } from "@/style";
import { mergeDeepRight } from "ramda";

export default plugin.withOptions<{
  className?: string;
}>(
  ({ className = "prose" } = {}) => {
    return ({ addComponents, theme }) => {
      const modifiers = theme("typography", {});

      addComponents(
        Object.entries(mergeDeepRight(styles, modifiers)).map(
          ([key, value]) => {
            return {
              [`.${className}${key === "DEFAULT" ? "" : `-${key}`}`]: value,
            } as { [key: string]: any };
          },
        ),
      );
    };
  },
  () => ({
    theme: { typography: styles },
  }),
);
