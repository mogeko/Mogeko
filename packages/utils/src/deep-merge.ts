import { deepMergeWith } from "@/deep-merge-with";

/**
 * Creates a new object with the own properties of the first object merged with
 * the own properties of the second object. If a key exists in all objects:
 *
 * - and all values are objects, the two values will be recursively merged
 * - otherwise the value from the second object will be used.
 *
 * @param objs - The objects to merge
 * @return The new object
 *
 * @example
 * ```typescript
 * deepMerge(
 *   { name: "fred", age: 10, contact: { email: "moo@example.com" } },
 *   { age: 40, contact: { email: "baa@example.com" } },
 * ); // { name: "fred", age: 40, contact: { email: "baa@example.com" } }
 * ```
 */
export const deepMerge = deepMergeWith((_a, b) => b);
