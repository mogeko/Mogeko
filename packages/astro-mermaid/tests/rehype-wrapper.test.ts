import { expect, describe, it } from "vitest";
import { wrapper } from "@/src/rehype-wrapper";
import { h } from "hastscript";

describe("wrapper", () => {
  it("Wrap mermaid code block with p tag", () => {
    const tree = h("body", [h("pre", { className: ["mermaid"] })]);

    wrapper()(tree);

    expect(tree).toMatchObject(
      h("body", [h("p", [h("pre", { className: ["mermaid"] })])]),
    );
    expect(tree).toMatchSnapshot();
  });

  it("Wrap mermaid code block with custom tag", () => {
    const tree = h("body", [h("pre", { className: ["mermaid"] })]);

    wrapper({ tagName: "div" })(tree);

    expect(tree).toMatchObject(
      h("body", [h("div", [h("pre", { className: ["mermaid"] })])]),
    );
    expect(tree).toMatchSnapshot();
  });

  it("Wrap pre with p tag", () => {
    const tree = h("pre", { className: ["mermaid"] });

    wrapper()(tree);

    expect(tree).toMatchObject(h("pre", { className: ["mermaid"] }));
    expect(tree).toMatchSnapshot();
  });

  it("Wrap multiple mermaid code block with p tag", () => {
    const tree = h("body", [
      h("pre", { className: ["mermaid"] }),
      h("pre", { className: ["mermaid"] }),
    ]);

    wrapper()(tree);

    expect(tree).toMatchObject(
      h("body", [
        h("p", [h("pre", { className: ["mermaid"] })]),
        h("p", [h("pre", { className: ["mermaid"] })]),
      ]),
    );
    expect(tree).toMatchSnapshot();
  });
});
