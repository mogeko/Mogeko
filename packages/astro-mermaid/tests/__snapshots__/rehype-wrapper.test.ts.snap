// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`wrapper > Wrap mermaid code block with custom tag 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "properties": {
            "className": [
              "mermaid",
            ],
          },
          "tagName": "pre",
          "type": "element",
        },
      ],
      "properties": {},
      "tagName": "div",
      "type": "element",
    },
  ],
  "properties": {},
  "tagName": "body",
  "type": "element",
}
`;

exports[`wrapper > Wrap mermaid code block with p tag 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "properties": {
            "className": [
              "mermaid",
            ],
          },
          "tagName": "pre",
          "type": "element",
        },
      ],
      "properties": {},
      "tagName": "p",
      "type": "element",
    },
  ],
  "properties": {},
  "tagName": "body",
  "type": "element",
}
`;

exports[`wrapper > Wrap multiple mermaid code block with p tag 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "properties": {
            "className": [
              "mermaid",
            ],
          },
          "tagName": "pre",
          "type": "element",
        },
      ],
      "properties": {},
      "tagName": "p",
      "type": "element",
    },
    {
      "children": [
        {
          "children": [],
          "properties": {
            "className": [
              "mermaid",
            ],
          },
          "tagName": "pre",
          "type": "element",
        },
      ],
      "properties": {},
      "tagName": "p",
      "type": "element",
    },
  ],
  "properties": {},
  "tagName": "body",
  "type": "element",
}
`;

exports[`wrapper > Wrap pre with p tag 1`] = `
{
  "children": [],
  "properties": {
    "className": [
      "mermaid",
    ],
  },
  "tagName": "pre",
  "type": "element",
}
`;
